mfhdf/ncdump/ncdump.1
.TH NCDUMP 1 "$Date$" "Printed: \n(yr-\n(mo-\n(dy" "UNIDATA UTILITIES"
.SH NAME
ncdump \- Convert netCDF files to ASCII form (CDL)
.SH SYNOPSIS
.ft B
.HP
ncdump
.nh
\%[-c]
\%[-h]
\%[-v \fIvar1,...\fP]
\%[-b \fIlang\fP]
\%[-f \fIlang\fP]
\%[-l \fIlen\fP]
\%[-n \fIname\fP]
\%[-d \fIf_digits[,d_digits]\fP]
\%\fIfile\fP
.hy
.ft
.SH DESCRIPTION
\fBncdump\fP generates an ASCII representation of a specified netCDF file on
standard output.  The ASCII representation is in a form called CDL
(``network Common Data form Language'') that can be viewed, edited, or serve
as input to \fBncgen\fP.  \fBncgen\fP is a companion program that can
generate a binary netCDF file from a CDL file.  Hence \fBncgen\fP and
\fBncdump\fP can be used as inverses to transform the data representation
between binary and ASCII representations.  See \fBncgen\fP for a description
of CDL and netCDF representations.
.LP
\fBncdump\fP defines a default format used for each type of netCDF data, but
this can be changed if a `C_format' attribute is defined for a netCDF
variable.  In this case, \fBncdump\fP will use the `C_format' attribute to
format each value.  For example, if floating-point data for the netCDF
variable `Z' is known to be accurate to only three significant digits, it
would be appropriate to use the variable attribute
.RS
.HP
Z:C_format = "%.3g"
.RE
.LP
\fBncdump\fP may also be used as a simple browser for netCDF data
files, to display the dimension names and sizes; variable names, types,
and shapes; attribute names and values; and optionally, the values of
data for all variables or selected variables in a netCDF file.
.SH OPTIONS
.IP "\fB-c\fP"
Show the values of \fIcoordinate\fP variables (variables that are also
dimensions) as well as the declarations of all dimensions, variables, and
attribute values.  Data values of non-coordinate variables are not included
in the output.  This is the most suitable option to use for a brief look at
the structure and contents of a netCDF file.
.IP "\fB-h\fP"
Show only the \fIheader\fP information in the output, that is the
declarations of dimensions, variables, and attributes but no data values for
any variables.  The output is identical to using the \fB-c\fP option except
that the values of coordinate variables are not included.  (At most one of
\fB-c\fP or \fB-h\fP options may be present.)
.IP "\fB-v\fP \fIvar1,...,varn\fP"
The output will include data values for the specified variables, in addition
to the declarations of all dimensions, variables, and attributes.  One or
more variables must be specified by name in the comma-delimited list
following this option.  The list must be a single argument to the command,
hence cannot contain blanks or other white space characters.  The named
variables must be valid netCDF variables in the input-file.  The default,
without this option and in the absence of the \fB-c\fP or \fB-h\fP
options, is to include data values for \fIall\fP variables in the output.
.IP "\fB-b\fP \fIlang\fP"
A brief annotation in the form of a CDL comment (text beginning with the
characters ``//'') will be included in the data section of the output for
each `row' of data, to help identify data values for multidimensional
variables.  If \fIlang\fP begins with `C' or `c', then C language
conventions will be used (zero-based indices, last dimension varying
fastest).  If \fIlang\fP begins with `F' or `f', then Fortran language
conventions will be used (one-based indices, first dimension varying
fastest).  In either case, the data will be presented in the same order;
only the annotations will differ.  This option is useful for browsing
through large volumes of multidimensional data.
.IP "\fB-f\fP \fIlang\fP"
Full annotations in the form of trailing CDL comments (text beginning with
the characters ``//'') for every data value (except individual characters in
character arrays) will be included in the data section.  If \fIlang\fP
begins with `C' or `c', then C language conventions will be used (zero-based
indices, last dimension varying fastest).  If \fIlang\fP begins with `F' or
`f', then Fortran language conventions will be used (one-based indices,
first dimension varying fastest).  In either case, the data will be
presented in the same order; only the annotations will differ.  This option
may be useful for piping data into other filters, since each data value
appears on a separate line, fully identified.
.IP "\fB-l\fP \fIlen\fP"
Changes the default maximum line length (80) used in formatting lists of
non-character data values.
.IP "\fB-n\fP \fIname\fP"
CDL requires a name for a netCDF data set, for use by \fBncgen -b\fP in
generating a default netCDF file name.  By default, \fIncdump\fP constructs
this name from the last component of the pathname of the input netCDF file
by stripping off any extension it has.  Use the \fB-n\fP option to specify a
different name.  Although the output file name used by \fBncgen -b\fP can be
specified, it may be wise to have \fIncdump\fP change the default name to
avoid inadvertently overwriting a valuable netCDF file when using
\fBncdump\fP, editing the resulting CDL file, and using \fBncgen -b\fP to
generate a new netCDF file from the edited CDL file.
.IP "\fB-d\fP \fIfloat_digits[,double_digits]\fP"
Specifies default number of significant digits to use in displaying
floating-point or double precision data values for variables that don't have
a `C_format' attribute.  Floating-point data will be displayed with
\fIfloat_digits\fP significant digits.  If \fIdouble_digits\fP is also
specified, double-precision values will be displayed with that many
significant digits.  If a variable has a `C_format' attribute, that
overrides any specified floating-point default.  In the absence of any
\fB-d\fP specifications, floating-point and double-precision data are
displayed with 7 and 15 significant digits respectively.  CDL files can be
made smaller if less precision is required.  If both floating-point and
double-presision precisions are specified, the two values must appear
separated by a comma (no blanks) as a single argument to the command.
If you really want every last bit of precision from the netCDF file
represented in the CDL file for all possible floating-point values, you will
have to specify this with \fB-d 9,17\fP (according to Theorem 15 of the
paper listed under REFERENCES).

.SH EXAMPLES
.LP
Look at the structure of the data in the netCDF file `\fBfoo.nc\fP':
.RS
.HP
ncdump -c foo.nc
.RE
.LP
Produce an annotated CDL version of the structure and data in the
netCDF file `\fBfoo.nc\fP', using C-style indexing for the annotations:
.RS
.HP
ncdump -b c foo.nc > foo.cdl
.RE
.LP
Output data for only the variables `uwind' and `vwind' from the netCDF file
`\fBfoo.nc\fP', and show the floating-point data with only three significant
digits of precision:
.RS
.HP
ncdump -v uwind,vwind -d 3 foo.nc
.RE
.LP
Produce a fully-annotated (one data value per line) listing of the data for
the variable `omega', using Fortran conventions for indices, and changing the
netCDF dataset name in the resulting CDL file to `omega':
.RS
.HP
ncdump -v omega -f fortran -n omega foo.nc > Z.cdl
.RE
.SH REFERENCES
 \fIWhat
Every Computer Scientist should Know About Floating-Point Arithmetic\fP, D.
Goldberg, \fBACM Computing Surveys, Vol. 23, No. 1\fP, March 1991, pp. 5-48.

.SH "SEE ALSO"
.LP
.BR ncgen (1),
.BR netcdf (3)
.SH BUGS
.LP
Character arrays that contain a null-byte are treated like C strings, so no
characters after the null byte appear in the output.

Multidimensional character string arrays are not handled well, since the CDL
syntax for breaking a long character string into several shorter lines is
weak.

There should be a way to specify that the data should be displayed in
`record' order, that is with the all the values for `record' variables
together that have the same value of the record dimension.


mfhdf/ncdump/ncdump.c
/*********************************************************************
 *   Copyright 1993, University Corporation for Atmospheric Research
 *   See netcdf/README file for copying and redistribution conditions.
 *********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include "local_nc.h"
#include "ncdump.h"

#include "h4getopt.h"
#include "dumplib.h"
#include "vardata.h"

static const char *type_name(nc_type type);

char        *progname;
struct ncdim dims[H4_MAX_NC_DIMS];  /* dimensions */
long         vdims[H4_MAX_NC_DIMS]; /* dimension sizes for a single variable */

static void
usage()
{
    printf("ncdump [-V|-c|-h|-u] [-v ...] [[-b|-f] [c|f]] [-l len] [-n name] [-d n[,n]] file\n");
    printf("\t [-V]             Display version of the HDF4 library and exit\n");
    printf("\t [-c]             Coordinate variable data and header information\n");
    printf("\t [-h]             Header information only, no data\n");
    printf("\t [-u]             Replace nonalpha-numerics in names with underscores\n");
    printf("\t [-v var1[,...]]  Data for variable(s) <var1>,... only\n");
    printf("\t [-b [c|f]]       Brief annotations for C or Fortran indices in data\n");
    printf("\t [-f [c|f]]       Full annotations for C or Fortran indices in data\n");
    printf("\t [-l len]         Line length maximum in data section (default 80)\n");
    printf("\t [-n name]        Name for netCDF (default derived from file name)\n");
    printf("\t [-d n[,n]]       Approximate floating-point values with less precision\n");
    printf("\t file             File name of input netCDF file\n");

    exit(EXIT_FAILURE);
}

/*
 * convert pathname of netcdf file into name for cdl unit, by taking
 * last component of path and stripping off any extension.
 */
static char *
name_path(char *path)
{
    char *cp, *new;

#ifdef MSDOS
#define FILE_DELIMITER '\\'
#endif
#ifndef FILE_DELIMITER /* default to unix */
#define FILE_DELIMITER '/'
#endif
    cp = strrchr(path, FILE_DELIMITER);
    if (cp == 0) /* no delimiter */
        cp = path;
    else /* skip delimeter */
        cp++;
    new = (char *)malloc((unsigned)(strlen(cp) + 1));
    if (new == 0) {
        error("out of memory!");
        exit(EXIT_FAILURE);
    }
    (void)strcpy(new, cp); /* copy last component of path */
    if ((cp = strrchr(new, '.')) != NULL)
        *cp = '\0'; /* strip off any extension */
    return new;
}

static const char *
type_name(nc_type type)
{
    switch (type) {
        case NC_BYTE:
            return "byte";
        case NC_CHAR:
            return "char";
        case NC_SHORT:
            return "short";
        case NC_LONG:
            return "long";
        case NC_FLOAT:
            return "float";
        case NC_DOUBLE:
            return "double";
        default:
            error("type_name: bad type %d", type);
            return "bogus";
    }
}

/*
 * Remove trailing zeros (after decimal point) but not trailing decimal
 * point from ss, a string representation of a floating-point number that
 * might include an exponent part.
 */
/* ss - returned string representing dd */
static void
tztrim(char *ss)
{
    char *cp, *ep;

    cp = ss;
    if (*cp == '-')
        cp++;
    while (isdigit((int)*cp) || *cp == '.')
        cp++;
    if (*--cp == '.')
        return;
    ep = cp + 1;
    while (*cp == '0')
        cp--;
    cp++;
    if (cp == ep)
        return;
    while (*ep)
        *cp++ = *ep++;
    *cp = '\0';
    return;
}

/*
 * Print list of attribute values.  Attribute values must be printed with
 * explicit type tags, because their types are not declared.
 */
static void
pr_att_vals(nc_type type, int len, void *vals)
{
    int iel;
    union {
        char   *cp;
        short  *sp;
        nclong *lp;
        float  *fp;
        double *dp;
    } gp;
    char         *sp;
    unsigned char uc;
    char          gps[30]; /* for ascii of a float or double precision */
    char         *f_fmt = "%#.8g";
    char         *d_fmt = "%#.16g";

    switch (type) {
        case NC_BYTE:
            gp.cp = (char *)vals;
            for (iel = 0; iel < len; iel++)
                if (isprint(uc = *gp.cp++ & 0377))
                    Printf("'%c'%s", uc, iel < len - 1 ? ", " : "");
                else
                    Printf("'\\%o'%s", uc, iel < len - 1 ? ", " : "");
            break;
        case NC_CHAR:
            gp.cp = (char *)vals;
            Printf("\"");
            /* adjust len so trailing nulls don't get printed */
            sp = gp.cp + len - 1;
            while (*sp-- == '\0' && len > 0)
                len--;
            for (iel = 0; iel < len; iel++)
                switch (uc = *gp.cp++ & 0377) {
                    case '\b':
                        Printf("\\b");
                        break;
                    case '\f':
                        Printf("\\f");
                        break;
                    case '\n': /* generate linebreaks after new-lines */
                        Printf("\\n\",\n    \"");
                        break;
                    case '\r':
                        Printf("\\r");
                        break;
                    case '\t':
                        Printf("\\t");
                        break;
                    case '\v':
                        Printf("\\v");
                        break;
                    case '\\':
                        Printf("\\\\");
                        break;
                    case '\'':
                        Printf("\\'");
                        break;
                    case '\"':
                        Printf("\\\"");
                        break;
                    default:
                        Printf("%c", uc);
                        break;
                }
            Printf("\"");
            break;
        case NC_SHORT:
            gp.sp = (short *)vals;
            for (iel = 0; iel < len; iel++)
                Printf("%ds%s", *gp.sp++, iel < len - 1 ? ", " : "");
            break;
        case NC_LONG:
            gp.lp = (nclong *)vals;
            for (iel = 0; iel < len; iel++)
                Printf("%d%s", (int)*gp.lp++, iel < len - 1 ? ", " : "");
            break;
        case NC_FLOAT:
            gp.fp = (float *)vals;
            for (iel = 0; iel < len; iel++) {
                int ll;
                (void)sprintf(gps, f_fmt, *gp.fp++);
                /* append a trailing "f" for floating-point attributes */
                ll          = strlen(gps);
                gps[ll + 1] = '\0';
                gps[ll]     = 'f';
                tztrim(gps); /* trim trailing 0's after '.' */
                Printf("%s%s", gps, iel < len - 1 ? ", " : "");
            }
            break;
        case NC_DOUBLE:
            gp.dp = (double *)vals;
            for (iel = 0; iel < len; iel++) {
                (void)sprintf(gps, d_fmt, *gp.dp++);
                tztrim(gps); /* trim trailing 0's after '.' */
                Printf("%s%s", gps, iel < len - 1 ? ", " : "");
            }
            break;
        default:
            error("pr_att_vals: bad type");
    }
}

/*
 * sanitize_string
 *
 * 	If the string contains characters other than alpha-numerics,
 * 	an underscore, or a hyphen, convert it to an underscore.
 */
char *
sanitize_string(char *str, bool sanitize)
{
    char *new_str = NULL;

    if (!str)
        return NULL;

    new_str = HDstrdup(str);
    if (NULL == new_str) {
        error("Out of memory!");
        return NULL;
    }

    if (sanitize) {
        char *ptr = new_str;

        /* Convert all non-alpha, non-numeric, non-hyphen, non-underscore
         * characters to underscores
         */
        for (; *ptr; ptr++)
            if (!isalnum(*ptr) && *ptr != '_' && *ptr != '-')
                *ptr = '_';
    }

    return new_str;
}

static void
do_ncdump(char *path, struct fspec *specp)
{
    int          ndims;       /* number of dimensions */
    int          nvars;       /* number of variables */
    int          ngatts;      /* number of global attributes */
    int          xdimid;      /* id of unlimited dimension */
    int          dimid;       /* dimension id */
    int          varid;       /* variable id */
    struct ncvar var;         /* variable */
    struct ncatt att;         /* attribute */
    int          id;          /* dimension number per variable */
    int          ia;          /* attribute number */
    int          iv;          /* variable number */
    int          is_coord;    /* true if variable is a coordinate variable */
    int          isempty = 0; /* true if an old hdf dim has no scale values */

    int    ncid  = ncopen(path, NC_NOWRITE); /* netCDF id */
    vnode *vlist = newvlist();               /* list for vars specified with -v option */

    /* don't crash on error */
    ncopts = 0;

    if (ncid == -1) {
        error("ncopen failed on %s", path);
        return;
    }

    /*
     * If any vars were specified with -v option, get list of associated
     * variable ids
     */
    for (iv = 0; iv < specp->nlvars; iv++) {
        varid = ncvarid(ncid, specp->lvars[iv]);
        varadd(vlist, varid);
    }

    /* if name not specified, derive it from path */
    if (specp->name == NULL)
        specp->name = name_path(path);

    Printf("netcdf %s {\n", specp->name);

    /*
     * get number of dimensions, number of variables, number of global
     * atts, and dimension id of unlimited dimension, if any
     */
    (void)ncinquire(ncid, &ndims, &nvars, &ngatts, &xdimid);

    /* get dimension info */
    if (ndims > 0) {
        Printf("dimensions:\n");

        for (dimid = 0; dimid < ndims; dimid++) {
            char *fixed_str = NULL;

            if (ncdiminq(ncid, dimid, dims[dimid].name, &dims[dimid].size) < 0)
                fprintf(stderr, "Error calling ncdiminq on dimid = %d\n", dimid);

            fixed_str = sanitize_string(dims[dimid].name, specp->fix_str);
            if (fixed_str == NULL) {
                (void)ncclose(ncid);
                return;
            }

            if (dimid == xdimid)
                Printf("\t%s = %s ; // (%d currently)\n", fixed_str, "UNLIMITED", (int)dims[dimid].size);
            else
                Printf("\t%s = %ld ;\n", fixed_str, dims[dimid].size);

            free(fixed_str);
        }
    }

    Printf("\nvariables:\n");

    /* get variable info, with variable attributes */
    for (varid = 0; varid < nvars; varid++) {
        char *fixed_var;

        (void)ncvarinq(ncid, varid, var.name, &var.type, &var.ndims, var.dims, &var.natts);
        fixed_var = sanitize_string(var.name, specp->fix_str);

        if (fixed_var == NULL) {
            (void)ncclose(ncid);
            return;
        }

        Printf("\t%s %s", type_name(var.type), fixed_var);

        if (var.ndims > 0)
            Printf("(");

        for (id = 0; id < var.ndims; id++) {
            char *fixed_dim = sanitize_string(dims[var.dims[id]].name, specp->fix_str);

            if (fixed_dim == NULL) {
                (void)ncclose(ncid);
                free(fixed_var);
                return;
            }

            Printf("%s%s", fixed_dim, id < var.ndims - 1 ? ", " : ")");
            free(fixed_dim);
        }

        Printf(" ;\n");

        /* get variable attributes */
        for (ia = 0; ia < var.natts; ia++) {
            char *fixed_att;

            (void)ncattname(ncid, varid, ia, att.name);
            fixed_att = sanitize_string(att.name, specp->fix_str);

            if (!fixed_att) {
                (void)ncclose(ncid);
                free(fixed_var);
                return;
            }

            Printf("\t\t%s:%s = ", fixed_var, fixed_att);
            (void)ncattinq(ncid, varid, att.name, &att.type, &att.len);
            att.val = (void *)malloc((unsigned)att.len * nctypelen(att.type));

            if (!att.val) {
                error("Out of memory!");
                (void)ncclose(ncid);
                free(fixed_att);
                free(fixed_var);
                return;
            }

            (void)ncattget(ncid, varid, att.name, att.val);
            pr_att_vals(att.type, att.len, att.val);
            Printf(" ;\n");
            free(att.val);
            free(fixed_att);
        }

        free(fixed_var);
    }

    /* get global attributes */
    if (ngatts > 0)
        Printf("\n// global attributes:\n");

    for (ia = 0; ia < ngatts; ia++) {
        char *fixed_att;

        (void)ncattname(ncid, NC_GLOBAL, ia, att.name);
        fixed_att = sanitize_string(att.name, specp->fix_str);

        if (!fixed_att) {
            (void)ncclose(ncid);
            return;
        }

        Printf("\t\t:%s = ", fixed_att);

        (void)ncattinq(ncid, NC_GLOBAL, att.name, &att.type, &att.len);
        att.val = malloc((unsigned)(att.len * nctypelen(att.type)));

        if (!att.val) {
            error("Out of memory!");
            (void)ncclose(ncid);
            free(fixed_att);
            return;
        }

        (void)ncattget(ncid, NC_GLOBAL, att.name, att.val);
        pr_att_vals(att.type, att.len, att.val);
        Printf(" ;\n");
        free(att.val);
        free(fixed_att);
    }

    if (!specp->header_only) {
        if (nvars > 0)
            Printf("\ndata:\n");

        /* output variable data */
        for (varid = 0; varid < nvars; varid++) {
            /* if var list specified, test for membership */
            if (specp->nlvars > 0 && !varmember(vlist, varid))
                continue;

            (void)ncvarinq(ncid, varid, var.name, &var.type, &var.ndims, var.dims, &var.natts);

            if (specp->coord_vals) {
                /* Find out if this is a coordinate variable */
                is_coord = 0;

                for (dimid = 0; dimid < ndims; dimid++) {
                    if (strcmp(dims[dimid].name, var.name) == 0 && var.ndims == 1) {
                        is_coord = 1;
                        break;
                    }
                }

                if (!is_coord)
                    /* don't get data for non-coordinate vars */
                    continue;
            }

            /*
             * Only get data for variable if it is not a record variable,
             * or if it is a record variable and at least one record has
             * been written.
             */
#ifdef HDF
            /* skip the dimension vars which have dim strings only.  */
            {
                NC     *handle;
                NC_var *vp;

                isempty = 0;
                handle  = NC_check_id(ncid);

                if (handle->file_type == HDF_FILE) {
                    vp = NC_hlookupvar(handle, varid);
                    /* This is set up to take care of
                     * cases where an array has been
                     * defined but no data */
                    /* has yet been added. */
                    if ((vp->data_tag == DFTAG_SDS || vp->data_tag == DFTAG_SD) && (vp->data_ref == 0))
                        isempty = 1;
                }
            }

#endif /* HDF */
            if (isempty)
                continue;

            if (var.ndims == 0 || var.dims[0] != xdimid || dims[xdimid].size != 0) {
                int ret = 0;

                /* Collect variable's dim sizes */
                for (id = 0; id < var.ndims; id++)
                    vdims[id] = dims[var.dims[id]].size;

                /* if (vardata(&var, vdims, ncid, varid, specp) == -1)
                 */
                ret = vardata(&var, vdims, ncid, varid, specp);
                /* Keep the original failure detection until further improvement */
                if (ret == -1) {
                    error("can't output data for variable %s", var.name);
                    (void)ncclose(ncid);
                    return;
                }
                /* This failure indicates that vardata failed
                   because ncvarget fails.  (HDFFR-1468)
                   -BMR, 2015/01/20) */
                if (ret == ERR_READFAIL) {
                    error("Reading failed for variable %s, possibly the data is corrupted.", var.name);
                    (void)ncclose(ncid);
                    return;
                }
            }
        }
    }

    Printf("}\n");
    (void)ncclose(ncid);
}

static void
make_lvars(char *arg, struct fspec *fspecp)
{
    char  *cp    = arg;
    int    nvars = 1;
    char **cpp;

    /* compute number of variable names in comma-delimited list */
    fspecp->nlvars = 1;
    while (*cp++)
        if (*cp == ',')
            nvars++;

    fspecp->lvars = (char **)malloc(nvars * sizeof(char *));
    if (!fspecp->lvars) {
        error("out of memory");
        exit(EXIT_FAILURE);
    }

    cpp = fspecp->lvars;
    /* copy variable names into list */
    for (cp = strtok(arg, ","); cp != NULL; cp = strtok((char *)NULL, ",")) {

        *cpp = (char *)malloc(strlen(cp) + 1);
        if (!*cpp) {
            error("out of memory");
            exit(EXIT_FAILURE);
        }
        strcpy(*cpp, cp);
        cpp++;
    }
    fspecp->nlvars = nvars;
}

/*
 * Extract the significant-digits specifiers from the -d argument on the
 * command-line and update the default data formats appropriately.
 */
static void
set_sigdigs(char *arg)
{
    char *ptr        = arg;
    char *ptr2       = 0;
    long  flt_digits = 7;  /* default floating-point digits */
    long  dbl_digits = 15; /* default double-precision digits */
    char  flt_fmt[6];
    char  dbl_fmt[6];

    if (arg != 0 && strlen(arg) > 0 && arg[0] != ',')
        flt_digits = strtol(arg, &ptr, 10);

    if (flt_digits < 1 || flt_digits > 10) {
        error("unreasonable value for float significant digits: %d", flt_digits);
        exit(EXIT_FAILURE);
    }
    if (*ptr == ',')
        dbl_digits = strtol(ptr + 1, &ptr2, 10);
    if (ptr2 == ptr + 1 || dbl_digits < 1 || dbl_digits > 20) {
        error("unreasonable value for double significant digits: %d", dbl_digits);
        exit(EXIT_FAILURE);
    }
    (void)sprintf(flt_fmt, "%%.%dg", (int)flt_digits);
    (void)sprintf(dbl_fmt, "%%.%dg", (int)dbl_digits);
    set_formats(flt_fmt, dbl_fmt);
}

int
main(int argc, char *argv[])
{
    static struct fspec fspec = /* defaults, overridden on command line */
        {
            0,         /* construct netcdf name from file name */
            false,     /* print header info only, no data? */
            false,     /* just print coord vars? */
            false,     /* brief  comments in data section? */
            false,     /* full annotations in data section?  */
            false,     /* replace nonalpha-numeric with underscore?  */
            LANG_NONE, /* language conventions for indices */
            0,         /* if -v specified, number of variables */
            0          /* if -v specified, list of variable names */
        };
    int   c;
    int   i;
    char *ptr;
    int   max_len = 80; /* default maximum line length */

    h4opterr = 1;
    progname = argv[0];

    if (1 == argc) /* if no arguments given, print help and exit */
        usage();

    while ((c = h4getopt(argc, argv, "b:cf:hul:n:v:d:V")) != EOF)
        switch (c) {
            case 'V': /* display version of the library */
                printf("%s, %s\n\n", argv[0], LIBVER_STRING);
                exit(EXIT_SUCCESS);
            case 'h': /* dump header only, no data */
                fspec.header_only = true;
                break;
            case 'c': /* header, data only for coordinate dims */
                fspec.coord_vals = true;
                break;
            case 'n': /*
                       * provide different name than derived from
                       * file name
                       */
                fspec.name = h4optarg;
                break;
            case 'u': /* replace nonalpha-numerics with underscores */
                fspec.fix_str = true;
                break;
            case 'b': /* brief comments in data section */
                fspec.brief_data_cmnts = true;
                switch (tolower(h4optarg[0])) {
                    case 'c':
                        fspec.data_lang = LANG_C;
                        break;
                    case 'f':
                        fspec.data_lang = LANG_F;
                        break;
                    default:
                        error("invalid value for -b option: %s", h4optarg);
                        exit(EXIT_FAILURE);
                }
                break;
            case 'f': /* full comments in data section */
                fspec.full_data_cmnts = true;
                switch (tolower(h4optarg[0])) {
                    case 'c':
                        fspec.data_lang = LANG_C;
                        break;
                    case 'f':
                        fspec.data_lang = LANG_F;
                        break;
                    default:
                        error("invalid value for -b option: %s", h4optarg);
                        exit(EXIT_FAILURE);
                }
                break;
            case 'l': /* maximum line length */
                max_len = strtol(h4optarg, 0, 0);
                if (max_len < 10) {
                    error("unreasonably small line length specified: %d", max_len);
                    exit(EXIT_FAILURE);
                }
                break;
            case 'v': /* variable names */
                /* make list of names of variables specified */
                make_lvars(h4optarg, &fspec);
                break;
            case 'd': /* specify precision for floats */
                set_sigdigs(h4optarg);
                break;
            case '?':
                usage();
                break;
        }

    set_max_len(max_len);

    argc -= h4optind;
    argv += h4optind;

    i = 0;
    do {
        if (argc > 0)
            do_ncdump(argv[i], &fspec);
    } while (++i < argc);

    /* Release resources */
    free(fspec.name);
    for (i = 0, ptr = fspec.lvars; i < fspec.nlvars; i++, ptr++)
        free(ptr);
    free(fspec.lvars);

    return EXIT_SUCCESS;
}
